### PR TITLE
Option for O_SYNC writes for fs backend to protect against crashes

### PR DESCRIPTION
**Pull request by krishnasrinivas**
## Description
If MINIO_FS_OSYNC is enabled, writes for fs backend will be O_SYNC

## Motivation and Context
Customer requirement.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

File: cmd/config/errors.go [hunk 1]
```diff
@@ -24,6 +24,12 @@ var (
 		"Browser can only accept `on` and `off` values. To disable web browser access, set this value to `off`",
 	)
 
+	ErrInvalidFSOSyncValue = newErrFn(
+		"Invalid O_SYNC value",
+		"Please check the passed value",
+		"Can only accept `on` and `off` values. To enabe O_SYNC for fs backend, set this value to `on`",
```
Comment (harshavardhana) [inline cmd/config/errors.go]:
```suggestion
		"Can only accept `on` and `off` values. To enable O_SYNC for fs backend, set this value to `on`",
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Rest LGTM

**Review by kannappanr — APPROVED** [review]
LGTM

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#9d6a17f3-e0c1-4ec3-9c40-42a3a1e1046c) |
| mint-gateway-s3.sh | :x: [more...](#1763b857-4996-480f-b409-ec289f3d6f70) |
| mint-gateway-azure.sh | :x: [more...](#3efc2b48-c6ac-4fd5-aa68-7dd74517c9b6) |
<a name="3efc2b48-c6ac-4fd5-aa68-7dd74517c9b6"></a>
### 9581-212bd6d/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:30110
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp fb8ad8c432b6:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 5 minutes and 15 seconds
(4/15) Running aws-sdk-ruby tests ... done in 16 seconds
(5/15) Running awscli tests ... FAILED in 23 seconds
{
  "name": "awscli",
  "duration": 2143,
  "function": "aws --endpoint-url http://minio-dev4.minio.io:30110 s3api put-object --body /mint/data/datafile-0-b --bucket \nAn error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it. --key datafile-0-b\n",
  "status": "FAIL",
  "error": "usage: aws [options] <command> <subcommand> [<subcommand> ...] [parameters]To see help text, you can run:  aws help  aws <command> help  aws <command> <subcommand> helpUnknown options: occurred, (BucketAlreadyOwnedByYou), when, calling, the, CreateBucket, operation:, Your, previous, request, to, create, the, named, bucket, succeeded, and, you, already, own, it., error"
}
(5/15) Running healthcheck tests ... done in 1 seconds
(6/15) Running mc tests ... done in 2 minutes and 56 seconds
(7/15) Running minio-dotnet tests ... done in 1 minutes and 21 seconds
(8/15) Running minio-go tests ... done in 4 minutes and 9 seconds
(9/15) Running minio-java tests ... done in 26 minutes and 53 seconds
(10/15) Running minio-js tests ... done in 2 minutes and 46 seconds
(11/15) Running minio-py tests ... done in 9 minutes and 40 seconds
(12/15) Running s3cmd tests ... FAILED in 45 seconds
{
  "name": "s3cmd",
  "duration": "1701",
  "function": "test_copy_object",
  "status": "FAIL",
  "error": ""
}
(12/15) Running s3select tests ... done in 21 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
<a name="1763b857-4996-480f-b409-ec289f3d6f70"></a>
### 9581-212bd6d/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev3.minio.io:32347
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f3129be9c142:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... FAILED in 20 seconds
{
  "name": "awscli",
  "duration": 4220,
  "function": "aws --endpoint-url http://minio-dev3.minio.io:32347 s3api get-object --bucket awscli-mint-test-bucket-17484 --key awscli-mint-test-bucket-17484-object /tmp/datafile-0-b\n",
  "status": "FAIL",
  "error": "Etag mismatch for multipart 0 byte object"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 17 seconds
(7/15) Running minio-dotnet tests ... done in 28 seconds
(8/15) Running minio-go tests ... done in 29 seconds
(9/15) Running minio-java tests ... done in 24 seconds
(10/15) Running minio-js tests ... done in 33 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 1 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 4 seconds
(13/15) Running s3select tests ... done in 1 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="9d6a17f3-e0c1-4ec3-9c40-42a3a1e1046c"></a>
### 9581-212bd6d/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:30540
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b69e9d2a6256:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 46 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... done in 2 minutes and 20 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 43 seconds
(8/15) Running minio-dotnet tests ... done in 44 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 24 seconds
(10/15) Running minio-java tests ... done in 7 minutes and 13 seconds
(11/15) Running minio-js tests ... FAILED in 17 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 26,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}
(11/15) Running minio-py tests ... done in 4 minutes and 15 seconds
(12/15) Running s3cmd tests ... done in 29 seconds
(13/15) Running s3select tests ... done in 5 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :x: [more...](#f7784526-af09-4b0d-af86-368a572f77d4) |
| mint-gateway-azure.sh | :x: [more...](#7263017d-1ee5-4cf8-ba91-e67435e9e552) |
<a name="7263017d-1ee5-4cf8-ba91-e67435e9e552"></a>
### 9581-75b30ad/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev3.minio.io:30108
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b4c4cc267334:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 5 minutes and 28 seconds
(4/15) Running aws-sdk-ruby tests ... done in 15 seconds
(5/15) Running awscli tests ... done in 2 minutes and 18 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 1 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 21 seconds
(9/15) Running minio-go tests ... done in 3 minutes and 47 seconds
(10/15) Running minio-java tests ... done in 26 minutes and 15 seconds
(11/15) Running minio-js tests ... done in 2 minutes and 35 seconds
(12/15) Running minio-py tests ... done in 9 minutes and 54 seconds
(13/15) Running s3cmd tests ... FAILED in 42 seconds
{
  "name": "s3cmd",
  "duration": "1691",
  "function": "test_copy_object",
  "status": "FAIL",
  "error": ""
}
(13/15) Running s3select tests ... done in 22 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="f7784526-af09-4b0d-af86-368a572f77d4"></a>
### 9581-75b30ad/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:31788
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ff985b0628ae:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 1 seconds
(5/15) Running awscli tests ... FAILED in 22 seconds
{
  "name": "awscli",
  "duration": 4611,
  "function": "aws --endpoint-url http://minio-c3.minio.io:31788 s3api get-object --bucket awscli-mint-test-bucket-7289 --key awscli-mint-test-bucket-7289-object /tmp/datafile-0-b\n",
  "status": "FAIL",
  "error": "Etag mismatch for multipart 0 byte object"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 17 seconds
(7/15) Running minio-dotnet tests ... done in 29 seconds
(8/15) Running minio-go tests ... done in 20 seconds
(9/15) Running minio-java tests ... done in 23 seconds
(10/15) Running minio-js tests ... done in 40 seconds
(11/15) Running minio-py tests ... done in 57 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 5 seconds
(13/15) Running s3select tests ... done in 0 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Pull request merged**